File: src/main/java/ru/mdashlw/rankedwho/updater/Updater.java
package ru.mdashlw.rankedwho.updater;

import net.minecraft.event.ClickEvent;
import net.minecraft.event.HoverEvent;
import net.minecraft.util.ChatComponentText;
import net.minecraft.util.ChatStyle;
import net.minecraft.util.EnumChatFormatting;
import net.minecraft.util.IChatComponent;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import org.kohsuke.github.*;
import ru.mdashlw.rankedwho.RankedWho;
import ru.mdashlw.rankedwho.base.VersionInfo;
import ru.mdashlw.rankedwho.events.PlayerConnectedToServerEvent;

import java.io.IOException;

public class Updater {
    private static final String USERNAME = "mdashlw";
    private static final String REPOSITORY = "ranked-who";

    private IChatComponent message;

    public void checkForUpdates() {
        String currentVersion = RankedWho.class.getPackage().getImplementationVersion();
        VersionInfo latestVersion = retrieveLatestVersion();

        System.out.println("[RankedWho] Current version: " + currentVersion);

        if (latestVersion == null) {
            System.out.println("[RankedWho] Unable to check for updates.");
            return;
        }

        String tag = latestVersion.getTag();
        String downloadUrl = latestVersion.getDownloadUrl();

        if (tag.equals(currentVersion)) {
            System.out.println("[RankedWho] The mod is up to date.");
            return;
        }

        message = new ChatComponentText("-----------------------------------------------------\n")
                .setChatStyle(new ChatStyle()
                        .setColor(EnumChatFormatting.BLUE))
                .appendSibling(new ChatComponentText("RankedWho ")
                        .setChatStyle(new ChatStyle()
                                .setColor(EnumChatFormatting.GOLD)))
                .appendSibling(new ChatComponentText("is out of date.\nYour version: ")
                        .setChatStyle(new ChatStyle()
                                .setColor(EnumChatFormatting.WHITE)))
                .appendSibling(new ChatComponentText(currentVersion)
                        .setChatStyle(new ChatStyle()
                                .setColor(EnumChatFormatting.RED)))
                .appendSibling(new ChatComponentText(" | ")
                        .setChatStyle(new ChatStyle()
                                .setColor(EnumChatFormatting.GRAY)))
                .appendSibling(new ChatComponentText("Latest version: ")
                        .setChatStyle(new ChatStyle()
                                .setColor(EnumChatFormatting.WHITE)))
                .appendSibling(new ChatComponentText(tag)
                        .setChatStyle(new ChatStyle()
                                .setColor(EnumChatFormatting.GREEN)))
                .appendSibling(new ChatComponentText("\n\nDownload link:\n")
                        .setChatStyle(new ChatStyle()
                                .setColor(EnumChatFormatting.GOLD)
                                .setBold(true)))
                .appendSibling(new ChatComponentText(downloadUrl)
                        .setChatStyle(new ChatStyle()
                                .setColor(EnumChatFormatting.AQUA)
                                .setChatHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, new ChatComponentText("Click to open!")))
                                .setChatClickEvent(new ClickEvent(ClickEvent.Action.OPEN_URL, downloadUrl))))
                .appendSibling(new ChatComponentText("\n-----------------------------------------------------")
                        .setChatStyle(new ChatStyle()
                                .setColor(EnumChatFormatting.BLUE)));
        MinecraftForge.EVENT_BUS.register(this);
    }

    private VersionInfo retrieveLatestVersion() {
        try {
            GitHub gitHub = GitHub.connectAnonymously();
            GHUser user = gitHub.getUser(USERNAME);
            GHRepository repository = user.getRepository(REPOSITORY);
            GHRelease latestRelease = repository.getLatestRelease();

            if (latestRelease == null) {
                return null;
            }

            String tag = latestRelease.getTagName();
            String downloadUrl = null;

            for (GHAsset asset : latestRelease.getAssets()) {
                if (asset.getName().endsWith(".jar")) {
                    downloadUrl = asset.getBrowserDownloadUrl();
                }
            }

            return new VersionInfo(tag, downloadUrl);
        } catch (IOException exception) {
            exception.printStackTrace();
            return null;
        }
    }

    @SubscribeEvent
    public void onPlayerConnectedToServer(PlayerConnectedToServerEvent event) {
        if (message == null) {
            return;
        }

        event.getPlayer().addChatMessage(message);
    }
}

File: src/main/kotlin/ru/mdashlw/rankedwho/updater/Updater.kt
package ru.mdashlw.rankedwho.updater

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import net.minecraft.util.IChatComponent
import net.minecraftforge.common.MinecraftForge
import org.kohsuke.github.GitHub
import ru.mdashlw.rankedwho.RankedWho
import ru.mdashlw.rankedwho.util.toChatComponent

object Updater {
    private const val USERNAME = "mdashlw"
    private const val REPOSITORY = "ranked-who"

    private lateinit var currentVersion: String
    private lateinit var latestVersion: String

    // Don't use appendln as it appends \r\n on Windows
    val message: IChatComponent by lazy {
        buildString {
            append("§9§m--------------------------------------------------\n")
            append("§6RankedWho §fis out of date.\n")
            append("§fYour version: §c$currentVersion §7| §fLatest version: §a$latestVersion\n")
            append("\n")
            append("§6§lDownload from GitHub\n")
            append("§bhttps://github.com/$USERNAME/$REPOSITORY\n")
            append("§9§m--------------------------------------------------")
        }.toChatComponent()
    }

    fun init() {
        GlobalScope.launch {
            currentVersion = RankedWho::class.java.`package`.implementationVersion
            latestVersion = retrieveLatestVersion()
                ?: throw IllegalStateException("Unable to check for updates")

            if (currentVersion == latestVersion) {
                println("RankedWho is up to date.")
                return@launch
            }

            println("RankedWho is out of date, latest version: $latestVersion.")

            MinecraftForge.EVENT_BUS.register(Listener)
        }
    }

    private fun retrieveLatestVersion(): String? {
        val gh = GitHub.connectAnonymously()
        val ghUser = gh.getUser(USERNAME)
        val ghRepository = ghUser.getRepository(REPOSITORY)
        val ghRelease = ghRepository.latestRelease

        return ghRelease?.tagName
    }
}
